File: LeetCode/2021春季杯/3. 魔塔游戏.py
import heapq
class Solution:
    def magicTower(self, nums: List[int]) -> int:
        ssum = sum(nums)
        if ssum < 0:
            return -1
        ssum = 0
        hq = []
        res = 0
        for i in nums:
            ssum += i
            if i < 0:
                heapq.heappush(hq, i)
                if ssum < 0:
                    ssum -= heapq.heappop(hq)
                    res += 1
        return res
                

File: LeetCode/2021春季杯/3.魔塔游戏.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int magicTower(vector<int>& nums) {
        long long ssum{0};
        int res{0};
        for (int i : nums) ssum += i;
        if (ssum < 0) return -1;
        ssum = 0;
        priority_queue<int> pq;
        for (int i : nums) {
            ssum += i;
            if (i < 0) {
                pq.push(i);
                if (ssum < 0) {
                    ssum -= pq.top();
                    pq.pop();
                    res++;
                }
            }
        }
        return res;
    }
};